apm-agent-plugins/apm-rabbitmq/apm-rabbitmq-spring/src/main/java/co/elastic/apm/agent/rabbitmq/SpringAmqpBatchMessageListenerInstrumentation.java

```diff
@@ -47,21 +57,73 @@ public String getAdviceClassName() {
     }
 
     public static class MessageListenerContainerWrappingAdvice extends BaseAdvice {
-        protected static final MessageBatchHelper messageBatchHelper;
+        private static final Logger oneTimeTransactionCreationWarningLogger;
+        private static final MessageBatchHelper messageBatchHelper;
+        private static final MessagingConfiguration messagingConfiguration;
 
         static {
             ElasticApmTracer elasticApmTracer = GlobalTracer.requireTracerImpl();
             messageBatchHelper = new MessageBatchHelper(elasticApmTracer, transactionHelper);
+            messagingConfiguration = elasticApmTracer.getConfig(MessagingConfiguration.class);
+            oneTimeTransactionCreationWarningLogger = LoggerUtils.logOnce(LoggerFactory.getLogger("Spring-AMQP-Batch-Logger"));
         }
 
-        @Nullable
-        @Advice.AssignReturned.ToArguments(@ToArgument(0))
+        @Advice.AssignReturned.ToArguments(@ToArgument(index = 0, value = 0, typing = DYNAMIC))
         @Advice.OnMethodEnter(inline = false)
-        public static List<Message> beforeOnBatch(@Advice.Argument(0) @Nullable final List<Message> messageBatch) {
-            if (!tracer.isRunning() || tracer.currentTransaction() != null || messageBatch == null) {
-                return messageBatch;
+        public static Object[] beforeOnBatch(@Advice.This Object thiz,
+                                             @Advice.Argument(0) @Nullable final List<Message> messageBatch) {
+
+            List<Message> processedBatch = messageBatch;
+            Transaction batchTransaction = null;
+
+            if (tracer.isRunning() && messageBatch != null && !messageBatch.isEmpty()) {
+                AbstractSpan<?> active = tracer.getActive();
+                if (active == null && messagingConfiguration.getMessageBatchStrategy() == MessagingConfiguration.BatchStrategy.BATCH_HANDLING) {
+                    batchTransaction = tracer.startRootTransaction(thiz.getClass().getClassLoader());
+                    if (batchTransaction == null) {
+                        oneTimeTransactionCreationWarningLogger.warn("Failed to start Spring AMQP transaction for batch processing");
+                    } else {
+                        batchTransaction.withType("messaging")
+                            .withName("Spring AMQP Message Batch Processing")
+                            .activate();
+                    }
+                } else {
+                    oneTimeTransactionCreationWarningLogger.warn("Unexpected active span during batch message processing start: {}",
+                        active);
+                }
+
+                active = tracer.getActive();
```

Comment:
Isn't the `active` variable already set with the active span/transaction on line 80 ?

Reply:
Yes, but line 88 potentially activates a transaction. This basically says: "if there was an active span before, **or if we just created one and activated it**, I want to add span links in both cases".

Reply:
Because the flow of this method is a bit complicated and because the return value is complex one that may be affected by multiple decision branches, I prefer one return in such methods (related to the comment above), but it's definitely a mater of taste